posts/2017-10-02-serverless-vs-containers.md

```diff
@@ -0,0 +1,175 @@
+
+When serverless compute was first introduced in 2014, the workloads were pretty limited and focused around smaller jobs such as image/data manipulation. But then AWS introduced the API Gateway as an event source for Lambda functions.
+
+That changed everything. It became possible to create whole APIs that were powered by serverless compute. More and more AWS services integrated with the Lambda compute offering, making it possible to build even larger, more complex, fully serverless applications.
+
+But what is a serverless application, exactly? In sum, an architecture is serverless if it has these characteristics:
+
+- Event-driven workflow ("If X then Y")
+- Pay-per-execution
+- Zero administration
+- Auto-scaling
+- Short-lived, stateless functions
+
+### Advantages of serverless
+
+- Zero administration
+- Pay-per-execution
+- Zero cost for idle time
+- Auto-scaling
+- Faster time-to-market
+- Microservice nature —> Clear codebase separation
+- Significantly reduced administration and maintenance burden
+
+### Disadvantages of serverless
+
+- No standardization (though the CNCF is working on this)
+- Black box
+- Vendor lock-in
+- Cold starts
+- Complex apps can be hard to build
+
+## When to choose what?!
+
+Now that we've learned more about those two exciting new technologies it's time to ask ourselves:
+
+> "Which technology should I pick for my next project"?
+
+As with most technology-related choices the answer is: "It depends..."
+
+Containers are great if one needs the full flexibility to install and use software with specific version requirements.
+This flexibility ranges from the choice of the underlying operating system up to the full control of the installed
+programming language and runtime version.
+
+It's even possible to operate containers with different software stacks throughout a large container fleet. This makes it 
+especially interesting if an old, legacy system should be migrated into a containerized environment.
+
+Container scheduling and management systems such as [Kubernetes](https://kubernetes.io/) are the de-facto standard to manage
+large-scale container setups. It's one of the most active projects on GitHub and was started by Google employees who
+were heavily involved with the first iterations of Googles own homebrew container platform called "Bord". This means that
+best-practices are already built-in.
+
+However this flexibility comes with a price tag. To fully benefit from containers one needs to split up the monolithic
+application into separate microservices which in turn need to be rolled-out as individual groups of containers. Those
+containers need to communicate with each other which involves more tooling to be setup and operated. Other than that it's
+important to keep the containers and their operating systems up-to-date. This means that security fixes and other updates
```

Comment:
I'd also mention the general ops burden here -- you're likely going to need a larger ops team to handle a Kubernetes deployment.

Also, it could be overkill for a new startup to set up a full Kubernetes deployment when you're trying to move as quickly as possible.